src/QsCompiler/Transformations/ClassicallyControlledTransformation.cs

```diff
@@ -4,615 +4,472 @@
+                            else
+                            {
+                                controlOpInfo = BuiltIn.ApplyIfElseR;
+                            }
+
+                            var (zeroOpArg, oneOpArg) = (result == QsResult.Zero)
+                                ? (Utilities.CreateValueTupleExpression(condId, condArgs), Utilities.CreateValueTupleExpression(defaultId, defaultArgs))
+                                : (Utilities.CreateValueTupleExpression(defaultId, defaultArgs), Utilities.CreateValueTupleExpression(condId, condArgs));
+
+                            controlArgs = Utilities.CreateValueTupleExpression(conditionExpression, zeroOpArg, oneOpArg);
+
+                            targetArgs = ImmutableArray.Create(condArgs.ResolvedType, defaultArgs.ResolvedType);
                         }
-                        else if (ctl)
+                        else if (defaultScope == null)
                         {
-                            controlOpInfo = (result == QsResult.Zero)
-                            ? BuiltIn.ApplyIfZeroC
-                            : BuiltIn.ApplyIfOneC;
+                            if (adj && ctl)
+                            {
+                                controlOpInfo = (result == QsResult.Zero)
+                                ? BuiltIn.ApplyIfZeroCA
+                                : BuiltIn.ApplyIfOneCA;
+                            }
+                            else if (adj)
+                            {
+                                controlOpInfo = (result == QsResult.Zero)
+                                ? BuiltIn.ApplyIfZeroA
+                                : BuiltIn.ApplyIfOneA;
+                            }
+                            else if (ctl)
+                            {
+                                controlOpInfo = (result == QsResult.Zero)
+                                ? BuiltIn.ApplyIfZeroC
+                                : BuiltIn.ApplyIfOneC;
+                            }
+                            else
+                            {
+                                controlOpInfo = (result == QsResult.Zero)
+                                ? BuiltIn.ApplyIfZero
+                                : BuiltIn.ApplyIfOne;
+                            }
+
+                            controlArgs = Utilities.CreateValueTupleExpression(
+                                conditionExpression,
+                                Utilities.CreateValueTupleExpression(condId, condArgs));
+
+                            targetArgs = ImmutableArray.Create(condArgs.ResolvedType);
                         }
                         else
                         {
-                            controlOpInfo = (result == QsResult.Zero)
-                            ? BuiltIn.ApplyIfZero
-                            : BuiltIn.ApplyIfOne;
+                            return null; // ToDo: Diagnostic message - default body exists, but is not valid
```

Comment:
What does it mean for the body not to be valid? I don't understand what you mean with this comment.

Reply:
Do you mean the default block?

Reply:
Right, the body of the default block is not valid, but it exists. In this case we would not be able to do the conversion on either because we went with an all-or-nothing approach to converting an if-structure. If both the conditional body and the default body can't be converted, then we don't convert either of them. Here we need to raise a diagnostics message informing the user why this if-structure couldn't be converted.